packages/graphql/src/translate/create-relationship-validation-string.ts

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) "Neo4j"
+ * Neo4j Sweden AB [http://neo4j.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Node } from "../classes";
+import { RELATIONSHIP_REQUIREMENT_PREFIX } from "../constants";
+import { Context } from "../types";
+
+function createRelationshipValidationString({
+    node,
+    context,
+    varName,
+}: {
+    node: Node;
+    context: Context;
+    varName: string;
+}): string {
+    const nonNullRelationFields = node.relationFields.filter((field) => {
+        if (field.typeMeta.array) {
+            return false;
+        }
+
+        const isRequired = field.typeMeta.required;
+        const isUnionOrInterface = Boolean(field.union) || Boolean(field.interface);
+
+        return isRequired && !isUnionOrInterface;
+    });
+
+    if (!nonNullRelationFields.length) {
+        return "";
+    }
+
+    const nonNullPredicates = nonNullRelationFields.map((field) => {
+        const toNode = context.neoSchema.nodes.find((n) => n.name === field.typeMeta.name) as Node;
+        const inStr = field.direction === "IN" ? "<-" : "-";
+        const outStr = field.direction === "OUT" ? "->" : "-";
+        const relVarname = `${varName}_${field.fieldName}_${toNode.name}_unique`;
+
+        const subQuery = [
+            `CALL {`,
```

Comment:
**question:** Could there be issues by not prepending `WITH` before each subquery here, rather than prepending all relationship validation with a single `WITH`?

For instance, taking these type definitions:

```gql
type Director {
    id: ID!
}

type Genre {
    name: String!
}

type Movie {
    id: ID!
    director: Director! @relationship(type: "DIRECTED", direction: IN)
    genre: Genre! @relationship(type: "HAS_GENRE", direction: OUT)
}
```

And executing:

```gql
mutation {
  updateMovies(
    where: {id: "${movieId}"}
    disconnect: {director: {where: {node: {id: "${directorId1}"}}}}
    connect: {genre: {where: {node: {name: "${directorId2}"}}}}
  ) {
    movies {
      id
      director {
        id
      }
    }
  }
}
```

You get the following Cypher:

```
MATCH (this:Movie)
WHERE this.id = $this_id
WITH this
CALL {
        WITH this
        OPTIONAL MATCH (this_connect_genre0_node:Genre)
        WHERE this_connect_genre0_node.name = $this_connect_genre0_node_name
        FOREACH(_ IN CASE this WHEN NULL THEN [] ELSE [1] END | 
                FOREACH(_ IN CASE this_connect_genre0_node WHEN NULL THEN [] ELSE [1] END | 
                        MERGE (this)-[:HAS_GENRE]->(this_connect_genre0_node)
                )
        )
        RETURN count(*)
}
WITH this
CALL {
WITH this
OPTIONAL MATCH (this)<-[this_disconnect_director0_rel:DIRECTED]-(this_disconnect_director0:Director)
WHERE this_disconnect_director0.id = $updateMovies.args.disconnect.director.where.node.id
FOREACH(_ IN CASE this_disconnect_director0 WHEN NULL THEN [] ELSE [1] END | 
DELETE this_disconnect_director0_rel
)
RETURN count(*)
}
WITH this
CALL {
        WITH this
        MATCH (this)<-[this_director_Director_unique:DIRECTED]-(:Director)
        WITH count(this_director_Director_unique) as c
        CALL apoc.util.validate(NOT(c = 1), '@neo4j/graphql/RELATIONSHIP-REQUIREDMovie.director required', [0])
        RETURN c AS this_director_Director_unique_ignored
}
CALL {
        WITH this
        MATCH (this)-[this_genre_Genre_unique:HAS_GENRE]->(:Genre)
        WITH count(this_genre_Genre_unique) as c
        CALL apoc.util.validate(NOT(c = 1), '@neo4j/graphql/RELATIONSHIP-REQUIREDMovie.genre required', [0])
        RETURN c AS this_genre_Genre_unique_ignored
}
RETURN this { .id, director: head([ (this)<-[:DIRECTED]-(this_director:Director)   | this_director { .id } ]) } AS this
```

Note the following:

```
WITH this
CALL {
        WITH this
        MATCH (this)<-[this_director_Director_unique:DIRECTED]-(:Director)
        WITH count(this_director_Director_unique) as c
        CALL apoc.util.validate(NOT(c = 1), '@neo4j/graphql/RELATIONSHIP-REQUIREDMovie.director required', [0])
        RETURN c AS this_director_Director_unique_ignored
}
CALL {
        WITH this
        MATCH (this)-[this_genre_Genre_unique:HAS_GENRE]->(:Genre)
        WITH count(this_genre_Genre_unique) as c
        CALL apoc.util.validate(NOT(c = 1), '@neo4j/graphql/RELATIONSHIP-REQUIREDMovie.genre required', [0])
        RETURN c AS this_genre_Genre_unique_ignored
}
```

This has caused us issues in the past (I'm actually surprised this runs) because subqueries generally always need to be prepended with `WITH`. Potentially instead return a `WITH` here?

```suggestion
        const subQuery = [
            `WITH ${varName}`,
            `CALL {`,
```

Reply:
This will be fine if there is nothing in between the CALL statements!